### ISSUE TITLE
Connection via Kerberos does not work

### ISSUE THREAD

**Issue by ddeimeke** [SchoofsKelvin/vscode-sshfs#24]
I try to connect to a Linux host via Kerberos and it does not work.

The used PuTTY Session works well, but the connection via vscode-sshfs does not.

Maybe this is part of the problem:

```
Problems loading reference 'https://raw.githubusercontent.com/SchoofsKelvin/vscode-sshfs/27dec3277d9a50869c3d2a11db21da6e07a3223a/settingsschema.json': Unable to load schema from 'https://raw.githubusercontent.com/SchoofsKelvin/vscode-sshfs/27dec3277d9a50869c3d2a11db21da6e07a3223a/settingsschema.json'

Problems loading reference 'https://raw.githubusercontent.com/SchoofsKelvin/vscode-sshfs/27dec3277d9a50869c3d2a11db21da6e07a3223a/settingsschema.json': Unable to load schema from 'https://raw.githubusercontent.com/SchoofsKelvin/vscode-sshfs/27dec3277d9a50869c3d2a11db21da6e07a3223a/settingsschema.json'
```

As far as I can see it is blocked by our proxy because of "missing authorization" (should use Kerberos as well).

Access to the marketplace is working, I could donwload the referenced files via normal browser.

**Comment by SchoofsKelvin** [comment]
Those two files are only required for autocomplete and "syntax checking" for the configs. It shouldn't interfere with the actual SSH connection.

Does your PuTTY (session) have anything configured related to Kerberos? I assume `Connection > SSH > Auth > GSSAPI` has `Attempt GSSAPI authentation` enabled (by default) and that PuTTY handles connecting via Kerberos.

When the extension "uses" the PuTTY session, the only thing it does is copy (some) settings from it. It doesn't actually use PuTTY to connect.

**Comment by ddeimeke** [comment]
Yes, you are right. GSSAPI is defined in PuTTY sessions.

**Comment by SchoofsKelvin** [comment]
I've looked some into this, but right now, I can't think of a good way of solving this issue. Currently, these seem to be the options:
- I could install a kerberos package, but these require binaries, not easy to do with vscode extensions
- Install kerberos automatically or on demand, which would require `npm`/`yarn`, build tools, ...
- I'm able to start a shell over SSH using PuTTY's `plink`, but I can't use that for file transfer
- I might be able to write a wrapper for `psftp`, or use an existing wrapper

This is quite a big feature and might take a while. After looking into it, I don't expect this to be very easy to accomplish right now.

**Comment by SchoofsKelvin** [comment]
Note to self: mscdex/ssh2#333 and possibly [npm:kerberos](https://www.npmjs.com/package/kerberos)

**Comment by GitMensch** [comment]
>     * Install kerberos automatically or on demand, which would require `npm`/`yarn`, build tools, ...

Hm, wouldn't it be possible to just create the binaries and include those in either this or an additional extension, then try to delay-load those when GSSAPI is requested in the setup?

**Comment by SchoofsKelvin** [comment]
If you read the [package description](https://www.npmjs.com/package/kerberos), you'll see it's not that easy. While a single binary should work for Windows, it's a whole other story for Unix-based systems. _As for bundling or auto-installing build tools etc and build the first time the extension needs it, that's quite a big "feature" I'm not too keen about._ That's also disregarding the fact that I'd still have to change the ssh2 library to actually support Kerberos and make use of that package.

It would still be ideal if I could wrap a raw SSH connection (not a single TTY) for external programs. As far as I can tell, PuTTY 0.75 (and earlier?) might allow this based on their [changelog](https://www.chiark.greenend.org.uk/~sgtatham/putty/changes.html). This just seems to be a "client-side tool" they added which can interact with their "connection sharing IPC channels" which (on Windows) are available under `//./pipe/`, so I might even be able to make use of this connection sharing in PuTTY versions from even years ago. For Unix-based (or OpenSSH on Windows) I could perhaps use OpenSSH's [multiplexing](https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.mux), which seems to be its own protocol (while PuTTY uses "regular" SSH if you ignore the SSL layer and such). Quite a bit of research and prototyping required.

Basically, unless there's a (relatively) simple way of adding Kerberos to the ssh2 library itself, it seems unlikely to add support for it anytime soon. It's more likely (although not anytime soon) that I'd support attaching to multiplexed connections (and let PuTTY/OpenSSH deal with authentication).
